url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/util/MersenneTwisterFast.java

```diff
@@ -0,0 +1,1304 @@
+            z ^= (z << 15) & TEMPERING_MASK_C;      // TEMPERING_SHIFT_T(z)
+            z ^= (z >>> 18);                        // TEMPERING_SHIFT_L(z)
+
+            bits = (((((long) y) << 32) + (long) z) >>> 1);
+            val = bits % n;
+        } while (bits - val + (n - 1) < 0);
+        return val;
+    }
+
+    /** Returns a random double in the half-open range from [0.0,1.0).  Thus 0.0 is a valid
+    result but 1.0 is not. */
+    public final double nextDouble() {
+        int y;
+        int z;
+
+        if (mti >= N) // generate N words at one time
+        {
+            int kk;
+            final int[] mt = this.mt; // locals are slightly faster
+            final int[] mag01 = this.mag01; // locals are slightly faster
+
+            for (kk = 0; kk < N - M; kk++) {
+                y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + M] ^ (y >>> 1) ^ mag01[y & 0x1];
+            }
+            for (; kk < N - 1; kk++) {
+                y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + (M - N)] ^ (y >>> 1) ^ mag01[y & 0x1];
+            }
+            y = (mt[N - 1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
+            mt[N - 1] = mt[M - 1] ^ (y >>> 1) ^ mag01[y & 0x1];
+
+            mti = 0;
+        }
+
+        y = mt[mti++];
+        y ^= y >>> 11;                          // TEMPERING_SHIFT_U(y)
+        y ^= (y << 7) & TEMPERING_MASK_B;       // TEMPERING_SHIFT_S(y)
+        y ^= (y << 15) & TEMPERING_MASK_C;      // TEMPERING_SHIFT_T(y)
+        y ^= (y >>> 18);                        // TEMPERING_SHIFT_L(y)
+
+        if (mti >= N) // generate N words at one time
+        {
+            int kk;
+            final int[] mt = this.mt; // locals are slightly faster
+            final int[] mag01 = this.mag01; // locals are slightly faster
+
+            for (kk = 0; kk < N - M; kk++) {
+                z = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + M] ^ (z >>> 1) ^ mag01[z & 0x1];
+            }
+            for (; kk < N - 1; kk++) {
+                z = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
+                mt[kk] = mt[kk + (M - N)] ^ (z >>> 1) ^ mag01[z & 0x1];
+            }
+            z = (mt[N - 1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
+            mt[N - 1] = mt[M - 1] ^ (z >>> 1) ^ mag01[z & 0x1];
+
+            mti = 0;
+        }
+
+        z = mt[mti++];
+        z ^= z >>> 11;                          // TEMPERING_SHIFT_U(z)
+        z ^= (z << 7) & TEMPERING_MASK_B;       // TEMPERING_SHIFT_S(z)
+        z ^= (z << 15) & TEMPERING_MASK_C;      // TEMPERING_SHIFT_T(z)
+        z ^= (z >>> 18);                        // TEMPERING_SHIFT_L(z)
+
+        /* derived from nextDouble documentation in jdk 1.2 docs, see top */
+        return ((((long) (y >>> 6)) << 27) + (z >>> 5)) / (double) (1L << 53);
+    }
+
+    public final double nextGaussian() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 82 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.